(DOCSP-23425): JS automatic client reset with recovery (#2238)
mongodb/docs-realm: source/sdk/node/examples/reset-a-client-realm.txt

  The client first attempts to recover changes that have not yet synced.
  If the client cannot recover unsynced  data, it falls through to
  discard unsynced changes but continues to automatically perform the client reset.
  Choose this mode when you want to enable automatic client
  recovery to fall back to discard unsynced changes.
- :ref:`Discard unsynced changes mode <node-discard-unsynced-changes>`:
  Restores the realm to a syncable state by discarding changes made
  since the last sync.
- :ref:`Manual recovery mode <node-manually-recover-unsynced-changes>`:
  Downloads a new copy of the realm, and moves the unsyncable realm
  to a backup. Migrate unsynced data from the backup copy of the
  realm to the new syncable copy.

Automatic vs. Manual Client Reset
---------------------------------

The Realm SDKs provide client reset modes
that automatically handle most client reset errors.

Automatic client reset modes restore your local realm file to a syncable state without
closing the realm or missing notifications.
The following client reset modes support automatic client resets:

- Recover unsynced changes mode

> - Recover or discard unsynced Changes mode

Reviewer: ```suggestion
- Recover or discard unsynced changes mode
```

- Discard unsynced changes mode

The differences between these modes are based on how they handle
changes on the device that have not yet synced to the backend.
Only manual recovery mode does not perform an automatic client reset.

Choose recover unsynced changes mode to handle most client reset
scenarios automatically. This attempts to recover unsynced changes when a
client reset occurs.

If your app requires specific client reset logic that can't be handled automatically,
you may want or need to :ref:`add a manual client reset handler <node-manually-recover-unsynced-changes>`
to the automatic client reset mode.

.. _node-client-reset-recovery:

Client Reset with Recovery
--------------------------

.. versionadded:: 10.23.0

Client Recovery is a feature that is enabled by default when you configure Device Sync.
When Client Recovery is enabled, Realm Database automatically manages the

> client reset process in most cases. When you make schema changes the client

Reviewer 2: "When you make schema changes the client can recover unsynced changes when there are no schema changes." 

Did we mean for this to be something like "The client can recover unsynced changes when there are no schema changes, or non-breaking schema changes." Not sure how the "When you make schema changes" at the beginning of the sentence fits in.

can recover unsynced changes when there are no schema changes, or non-breaking schema changes.

> To use Client Recovery, configure your realm with recover unsynced changes or

Reviewer 2: I don't know how best to improve this, but this sentence reads weird to me - a bit like a run-on sentence. I wonder if we can put "recover unsynced changes" and "recover or discard unsynced changes" in monospace, or make them links, or what. Maybe italicize them? I feel like there should be some change in formatting to indicate that these are specific client reset modes we're discussing here. Maybe it's bullets? I.e.

> To use Client Recovery, configure your realm with one of these client reset modes:
> - recover unsynced changes
> - recover or discard unsynced changes

Or even something like:

> These client reset modes are compatible with Client Recovery:
> - recover unsynced changes
> - recover or discard unsynced changes

Author: i was thinking the same when applying the copy to Flutter PR. will adjust phrasing

recover or discard unsynced changes client reset modes.

.. include:: /includes/client-recovery-rules.rst

For more information about configuring Client Recovery, refer to
:ref:`Client Recovery <enable-or-disable-recovery-mode>` in the App Services documentation.

> Client Recovery cannot occur when your app makes breaking schema changes.

Reviewer: ```suggestion
Client Recovery cannot succeed when your app makes breaking schema changes.
```

A breaking change is a change that you can make in your server-side
schema that requires additional action to handle.

> In this scenario, client reset with fall back to a

Reviewer 2: I think "with" is either a typo for "will", or some words got left out. What about something like:

```suggestion
In this scenario, client reset falls back to a
```

:ref:`manual error client reset fallback <node-manual-client-reset-fallback>`.

For information on breaking vs. non-breaking schema changes, refer to
:ref:`breaking-change-quick-reference` in the App Services documentation.

.. _node-recover-unsynced-changes:

Recover Unsynced Changes Mode
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

When you choose **recover unsynced changes** mode,

> the client attempts to recover unsynced changes with Client Recovery.

Reviewer 2: This seemed like an unnecessary/redundant repeat of the phrase "recover unsynced changes." Kind of like using a word in its own dictionary definition. Maybe we could reword? What about something like:

> When you choose **recover unsynced changes** mode, the client attempts to integrate local changes that have not yet synced to the backend.

Choose this mode when you do not want to fall through to discard unsynced changes.

To handle client resets with the recover unsynced changes mode,
pass a :js-sdk:`ClientResetConfiguration <Realm.App.Sync.html#~ClientResetConfiguration>`
to the ``clientReset`` field of your
:js-sdk:`SyncConfiguration <Realm.App.Sync.html#~SyncConfiguration>`.
Include these properties in the ``ClientResetConfiguration``:

- ``mode``: Set to ``"recoverUnsyncedChanges"``.

> - ``clientResetBefore``: Optional. Callback function invoked before the SDK executes this mode,

Reviewer 2: Should these be ``onBefore`` and ``onAfter``?

Author: sure should! this is a change from the prev version which used `clientResetBefore` and clientResetAfter`.

  when the SDK receives a client reset error from the backend.
  Provides a copy of the realm.
- ``clientResetAfter``: Optional. Callback function invoked after the SDK successfully
  executes this mode. Provides instances of the realm before and after the client reset.
- ``onFallback``: Optional. Callback function which the SDK invokes
  only if the automatic recovery fails. For more information, refer to the
  :ref:`Manual Client Reset Fallback section <node-manual-client-reset-fallback>`.

The following example implements recover unsynced changes mode:

> .. literalinclude:: /examples/generated/node/client-reset.snippet.recover-unsynced-changes.js

Reviewer 2: I'm not sure what the convention is in the JS docs, so maybe this isn't consistent with how we generally do code examples here, but what about adding a comment to these `onBefore` and `onAfter` blocks with an example of the types of logic you might want to put here, instead of just a print statement? i.e.:

```js
 // This block could be used for custom recovery, reporting, debugging etc.
```

Author: historically, we do more console.logs, but i prefer your suggestion. will use here.

   :language: js

.. _node-recover-discard-unsynced-changes:

Recover or Discard Unsynced Changes Mode
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In **recover or discard unsynced changes** mode,
the client first attempts to recover changes that have not yet synced.
If the client cannot recover unsynced  data, it falls through to
discard unsynced changes but continues to automatically perform the client reset.
Choose this mode when you want to enable automatic client
recovery to fall back to discard unsynced changes.

To handle client resets with the recover or discard unsynced changes mode,
pass a :js-sdk:`ClientResetConfiguration <Realm.App.Sync.html#~ClientResetConfiguration>`
to the ``clientReset`` field of your
:js-sdk:`SyncConfiguration <Realm.App.Sync.html#~SyncConfiguration>`.
Include these properties in the ``ClientResetConfiguration``:

- ``mode``: Set to ``"recoverOrDiscardUnsyncedChanges"``.

> - ``clientResetBefore()``: Optional. Callback function invoked before the SDK executes this mode,

Reviewer 2: Same as above - wondering if these should be ``onBefore`` and ``onAfter``?

Also, curious why we include the ``()`` here but not above - we should probably be consistent about whether we include the parens across the doc.

Author: yep, good spot! i thought i'd gotten rid of this. looks like i missed this.

  For more information, refer to the :ref:`Manual Client Reset Fallback section
  <node-manual-client-reset-fallback>`.

The following example implements recover unsynced changes mode:

.. literalinclude:: /examples/generated/node/client-reset.snippet.recover-or-discard-unsynced-changes.js
   :language: js

.. _node-manual-client-reset-fallback:

Manual Client Reset Fallback
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If the client reset with recovery cannot complete automatically,
like when there are breaking schema changes, the client reset process falls through
to a manual error handler. This may occur in either of the client reset with recovery modes,
recover unsynced changes and recover or discard unsynced changes.

You must provide a manual client reset implementation
in the ``SyncConfiguration.onFallback()`` callback. ``onFallback()`` takes two
arguments:

- ``session``: :js-sdk:`Session <Realm.App.Sync.Session.html>` object representing
  the state of the Device Sync session.
- ``path``: String with the path to the current realm database file.

The following example demonstrates how you can manually handle this error case by
discarding all unsynced changes:

> .. literalinclude:: /examples/generated/node/client-reset.snippet.recovery-fallback.js

Reviewer 2: I really like this example - this is a great, practical flow for a good way to handle the manual recovery case.

Author: hat tip to .NET for this!

   :language: js

.. _node-discard-unsynced-changes:

Discard Unsynced Changes Mode
-----------------------------

.. versionadded:: 10.11.0

.. versionchanged:: 10.23.0 Mode renamed from "discardLocal" to "discardUnsyncedChanges". Both currently work, but in a future version, "discardLocal" will be removed. "clientResetBefore" and "clientResetAfter" callbacks renamed to "onBefore" and "onAfter", respectively.

**Discard Unsynced Changes** mode  permanently deletes all
local unsynced changes made since the last successful sync.
You might use this mode when your app requires client recovery logic that is not
consistent with the :ref:`automatic Client Recovery <node-client-reset-recovery>`,
or when you don't want to recover unsynced data.

> Do not use discard unsynced changes mode if your application cannot lose local

Reviewer 2: I think this sentence should also be in the "recover or discard" mode above, since that mode can fall back to discard and this is important in that scenario.

.. literalinclude:: /examples/generated/node/client-reset.snippet.discard-unsynced-changes-after-destructive-schema-changes.js
   :language: javascript

.. note:: Discard with Recovery

   If you'd like to attempt to recover unsynced changes, but but discard
   any changes that cannot be recovered, refer to the
   :ref:`recover or discard unsynced changes mode section <node-recover-discard-unsynced-changes>`.

.. _node-manually-recover-unsynced-changes:

Manual Mode
-----------

.. versionchanged:: 10.23.0 onManual callback added

> In **manual** mode, you define your own client reset handler.

Reviewer 2: I wonder if we want to say something here about why you might use manual mode? i.e. maybe our recovery rule logic doesn't work for your app and you can't discard unsynced local data?

To handle client resets with manual mode,
pass a :js-sdk:`ClientResetConfiguration <Realm.App.Sync.html#~ClientResetConfiguration>`
to the ``clientReset`` field of your
:js-sdk:`SyncConfiguration <Realm.App.Sync.html#~SyncConfiguration>`.
Include these properties in the ``ClientResetConfiguration``:

- ``mode``: Set to ``"manual"``.
- ``onManual``: Optional. Callback function invoked when the client reset occurs.
  Provides information about the sync session and the path to the current realm.
  If you don't set the ``onManual`` error handler, the client reset error falls
  back to the general sync error handler.

.. literalinclude:: /examples/generated/node/client-reset.snippet.manual-client-reset-onmanual.js
   :language: js

Manual Data Recovery
~~~~~~~~~~~~~~~~~~~~

To recover data from a manual client reset requires significant amounts
of code, schema concessions, and custom conflict resolution logic.

> To learn more about the **manually recover unsynced changes** client reset mode,

Reviewer 2: I don't think I'd call this "manually recover unsynced changes client reset mode." We don't really want devs to do this as it's a pretty tough problem and they're likely to have a bad time, and I think saying "recover unsynced changes" here implies some level of parity with our automatic recovery logic. We'd probably prefer to say something like:

> If you need to implement your own custom client reset logic, see the :ref:`Advanced Guide to...`
